test/core/event_engine/fuzzing_event_engine/fuzzing_event_engine.cc

```diff
@@ -534,5 +538,97 @@ FuzzingEventEngine::ListenerInfo::~ListenerInfo() {
       });
 }
 
+void FuzzingEventEngine::MockEndpoint::Init(
+    const MockEndpointActions& actions) {
+  Time last_action_exec_time = g_fuzzing_event_engine->Now();
+  for (auto action = actions.begin(); action != actions.end(); ++action) {
+    Time curr_action_exec_time =
+        std::get<Duration>(*action) + last_action_exec_time;
+    if (actions_by_time_.empty() ||
+        std::get<Time>(actions_by_time_.back()) != curr_action_exec_time) {
+      actions_by_time_.emplace_back(curr_action_exec_time,
+                                    std::get<std::string>(*action));
+    } else {
+      // Concatenate curr action with last action because the timestamps match
+      auto new_action =
+          absl::StrCat(std::get<std::string>(actions_by_time_.back()),
+                       std::get<std::string>(*action));
+      std::get<std::string>(actions_by_time_.back()) = new_action;
+    }
+    last_action_exec_time = curr_action_exec_time;
+  }
+}
+
+bool FuzzingEventEngine::MockEndpoint::Read(
+    absl::AnyInvocable<void(absl::Status)> on_read, SliceBuffer* buffer,
+    const ReadArgs* /*args*/) {
+  Duration delay = Duration::zero();
+  buffer->Clear();
+  if (!actions_by_time_.empty() &&
+      g_fuzzing_event_engine->Now() <
+          std::get<Time>(actions_by_time_.front())) {
+    delay = std::get<Time>(actions_by_time_.front()) -
+            g_fuzzing_event_engine->Now();
+  }
+  g_fuzzing_event_engine->RunAfter(delay, [on_readable = std::move(on_read),
+                                           buffer, this]() mutable {
+    auto tcp_annotate_error = [this](absl::Status src_error) {
+      auto peer_string = ResolvedAddressToNormalizedString(peer_address_);
+
+      grpc_core::StatusSetStr(&src_error,
+                              grpc_core::StatusStrProperty::kTargetAddress,
+                              peer_string.ok() ? *peer_string : "");
+      grpc_core::StatusSetInt(&src_error, grpc_core::StatusIntProperty::kFd,
+                              -1);
+      grpc_core::StatusSetInt(&src_error,
+                              grpc_core::StatusIntProperty::kRpcStatus,
+                              GRPC_STATUS_UNAVAILABLE);
+      return src_error;
+    };
+    // Treat the non existence of data available to read as an Endpoint
+    // error.
+    if (actions_by_time_.empty()) {
+      on_readable(tcp_annotate_error(absl::InternalError("Socket Closed")));
+    } else {
+      // Present the first set of bytes as the data read at this time.
+      do {
+        buffer->Append(Slice::FromCopiedString(
+            std::get<std::string>(actions_by_time_.front())));
+        actions_by_time_.erase(actions_by_time_.begin());
+      } while (!actions_by_time_.empty() &&
+               g_fuzzing_event_engine->Now() >=
+                   std::get<Time>(actions_by_time_.front()));
```

Comment:
This seems to imply that `actions_by_time_` is a serial of actions sorted by timepoints (earlier one happens first). But the data structure doesn't guarantee that.

Maybe a `std::map<Time, std::string> actions_by_time_;` could work? Then this part would become:
```
while (!actions_by_time_.empty() &&
       actions_by_time_.begin()->first <= g_fuzzing_event_engine->Now()) {
  buffer->Append(Slice::FromCopiedString(actions_by_time_.begin()->second));
  actions_by_time_.erase(actions_by_time_.begin());
}
```

Reply:
If this is doable, I believe the insertion and concatenation part of the code in `Init()` can be changed accordingly. Let me know what you think!

Reply:
Oh sorry, I think I missed the point that `actions_by_time_` **is** a serial of actions sorted by timepoints. Please feel free to disregard this comment!